### PR TITLE
Allow copying property path from inspector.

### LINKED ISSUE #106: Ability to copy parameters path from shaders and animation trees
**Describe the project you are working on:**
Currently i am developing a few small game-jam like projects

**Describe how this feature / enhancement will help your project:**
Not having to transcribe the values of paths will save some time and help avoid bugs when mistyping

**Show a mock up screenshots/video or a flow diagram explaining how your proposal will work:**
I hover mouse over the property with shader parameter
image
and now i want to press or do something that will save property path to clipboard so that i can easily copy it into the script

**Describe implementation detail for your proposal (in code), if possible:**
No clue, if someone could direct me somewhere i could do it myself probably, but right now i don't know much about godot to propose code

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
Not that i know of, if it is, i would love to make a plugin for it then

**Is there a reason why this should be core and not an add-on in the asset library?:**
Do not know if it is possible to do that, it might be because some inspector fields can tell the difference between right and left click

### PR DESCRIPTION
**Pull request by rcorre**
If you hover over a property label in the inspector dock and press
the "property_ediotr/copy_property_path" shortcut (ctrl+c by default),
the complete property path will be copied to the system clipboard. This
is especially useful for the `AnimationTree`, where code might reference
properties like "parameters/state/aim/move/blend_position".

Resolves godotengine/godot-proposals#106.

One issue is that if you click a property, then click on the node you
currently have selected in the node tree, then press ctrl+c, it will
still copy the selected property path rather than the node path. If you
click on a different node in the nodetree, however, ctrl+c will return
to copying the nodepath.

### CODE REVIEW

**Comment by KoBeWi** [comment]
There are 2 related PRs right now: #39398 and #34892

The reason I bring this up is that if the all 3 were going to be merged, the shortcut for your feature would need to changed to Ctrl + Shift + C for consistency (my scene tree PR changes Copy Node Path to this shortcut). But this is probably not important for now, as it's uncertain which of these PRs are going to make it 🤔

**Comment by rcorre** [comment]
@KoBeWi it seems like your PR adds `ctrl+c` as an [accelerator](https://docs.godotengine.org/en/latest/classes/class_popupmenu.html#class-popupmenu-method-add-item), not as a global keyboard shortcut. I just tried it out and it seems to work for me, as in `ctrl+c` will focus the menu item for copy, but does not automatically copy if the menu isn't open.

I'm thinking we should add two separate editor shortcuts, "property_editor/copy_property_path" and "property_editor/copy_property", so the user can create individual bindings for either, both separate from "copy_node_path".

**Comment by rcorre** [comment]
(BTW I'm trying to combine the two efforts now)

**Comment by rcorre** [comment]
NVM, I misread. Your interpretation of accelerators is correct.

**Comment by rcorre** [comment]
@KoBeWi I think I combined the two in a way that makes sense. Now the popup menu has options for "copy property (ctrl+c)", "paste property (ctrl+v)", and "copy property path (ctrl+shift+c)". The keybindings work -- I followed the same pattern that seemed to be used for other popup menus with shortcuts, e.g. https://github.com/rcorre/godot/blob/d110aa421b04312e428baf93e97f40049e452381/editor/scene_tree_dock.cpp#L2467.

**Comment by rcorre** [comment]
I'm getting these warnings though, which I haven't figured out, as `InspectorDock` seems to be constructed before `InspectorProperty`:

```
ERROR: Used ED_GET_SHORTCUT with invalid shortcut: property_editor/copy_property.
   at: ED_GET_SHORTCUT (editor/editor_settings.cpp:1514)
ERROR: Cannot add item with invalid ShortCut.
   at: add_shortcut (scene/gui/popup_menu.cpp:711)
ERROR: Used ED_GET_SHORTCUT with invalid shortcut: property_editor/paste_property.
   at: ED_GET_SHORTCUT (editor/editor_settings.cpp:1514)
ERROR: Cannot add item with invalid ShortCut.
```

Its also weird that it throws for those two, but not `copy_property_path`. The shortcuts seem to work in spite of the warnings.

**Comment by HEAVYPOLY** [comment]
Heads up this doesn't seem to be working on Mac with 3.2.3 rc3.  Also the editor hotkey for copy property was blank by default.  I tried setting to Command C and not getting any copy. 
image

**Comment by akien-mga** [comment]
@HEAVYPOLY This hasn't been merged yet so it's not included in 3.2.3 RC 3. Or do you mean that you cherry-picked this PR manually and made your own build?

**Comment by rcorre** [comment]
I just backported to 3.2 [here](https://github.com/rcorre/godot/commits/proposal-106-copy-prop-path-backport-3.2) and it seems to work. I don't have a mac to test on though.

**Comment by HEAVYPOLY** [comment]
@akien-mga Ah my mistake.  Look forward to trying this in the next release!

**Review by KoBeWi — APPROVED** [review]
Needs rebase, but otherwise looks good.

EDIT:
Well, seems like there is issue with Ctrl+V changing pivot if compatible node is selected (e.g. Sprite). The inspector doesn't correctly grab focus. This wasn't important before, but now that we have this inspector-specific shortcut, this should be solved somehow. Not sure how though, it could be done in another PR.

EDIT2:
Also you need to squash the commits and update the commit message.

**Comment by rcorre** [comment]
@KoBeWi I see what you're saying about the pivot shortcut, though it also conflicts with the existing copy/paste node shortcuts, so it seems like a more general problem that needs to be solved outside the context of this.

Is there a customizeable shortcut for changing the pivot? I couldn't find it (even typing `ctrl+v` into the editor shortcuts menu).

**Comment by KoBeWi** [comment]
> Is there a customizeable shortcut for changing the pivot? I couldn't find it (even typing ctrl+v into the editor shortcuts menu).

The shortcut is just V and it's not customizable. It wouldn't be a problem if clicking a property would focus the inspector.

**Review by KoBeWi — CHANGES_REQUESTED** [review]
Wait, actually there is one thing that needs to be changed. Inspector creates many EditorProperties at once and creating a menu for each of them is inefficient. They should be created on demand. See #50530 Just add a method `_ensure_popup()` and call it each time the popup is supposed to open.

Another thing which will be relevant after #51722 is that you need to block the popup and shortcuts when the property is readonly (by using `is_read_only()`).

**Comment by rcorre** [comment]
The popup is now created just-in-time. I'm assuming that if it is readonly, I should block the paste shortcut, but not the popup as a whole, since copy/copy_path should still be allowed, right?

File: editor/editor_inspector.cpp [hunk 1]
```diff
@@ -784,6 +786,30 @@ void EditorProperty::_gui_input(const Ref<InputEvent> &p_event) {
 			update();
 			emit_signal(SNAME("property_checked"), property, checked);
 		}
+	} else if (mb.is_valid() && mb->is_pressed() && mb->get_button_index() == MOUSE_BUTTON_RIGHT) {
+		_ensure_popup();
+		menu->set_position(get_global_transform().xform(get_local_mouse_position()));
```
Comment (KoBeWi) [inline editor/editor_inspector.cpp:791]: This isn't correct way to set popup position anymore. Use `menu->set_position(get_screen_position() + get_local_mouse_position());` instead.

**Comment by KoBeWi** [comment]
You are right, only paste should be blocked.

btw I now noticed that shortcut checking in `_unhandled_key_input()` and switch cases in `menu_option()` are ordered differently than the enum. You should reorder the code.

**Comment by rcorre** [comment]
Fixed the enum ordering and respected is_read_only:

menu

**Review by KoBeWi — APPROVED** [review]
Looks good now.

Will need update if #51983 gets merged first.

**Comment by pycbouh** [comment]
I don't see it mentioned in the OP, but from the last screenshot... does this implement https://github.com/godotengine/godot-proposals/issues/3098?

**Comment by KoBeWi** [comment]
Partially, Open in Help would be missing.

**Comment by pycbouh** [comment]
Yeah, and reset I guess? But since the proposal is based around the problem of copying the path, I think we can say that it closes it, and maybe do a follow-up PR that adds extra options.

**Comment by KoBeWi** [comment]
I wouldn't close the proposal, rather just leave a comment that it's halfway there.

**Comment by akien-mga** [comment]
This would require a dedicated PR for `3.x` if a backport is wanted.

**Comment by rcorre** [comment]
Pretty sure I still have the old `3.x` branch around. I can work on a backport if we want one.

**Comment by Calinou** [comment]
> Pretty sure I still have the old `3.x` branch around. I can work on a backport if we want one.

A backport to `3.x` would be welcome (so that it can be merged for 3.5).

**Comment by Calinou** [comment]
I can't find the `3.x` version of this pull request. Can anyone link it here?

**Comment by KoBeWi** [comment]
#54913

**Pull request merged**